### ISSUE TITLE
Feature: Training data quantisation to prevent memorisation (#1899)

### ISSUE THREAD

**Issue by stservice** [stSoftwareAU/NEAT-TS#1901]
## Summary

Implement automatic quantisation of training data values during the training loop to reduce precision and prevent networks from memorising exact training examples. Quantisation rounds continuous values to a finite set of discrete levels, forcing the network to learn robust mappings rather than exact data points.

### 1. Create `DataQuantisationConfig.ts`

Create `src/config/DataQuantisationConfig.ts` following the established config pattern (`CrossValidationConfig.ts` is a good reference):

- `DataQuantisationConfig` — optional fields interface
- `RequiredDataQuantisationConfig` — `Required<DataQuantisationConfig>`
- `DEFAULT_DATA_QUANTISATION_CONFIG` — default values

**Configuration fields:**

| Field | Type | Default | Range | Description |
|-------|------|---------|-------|-------------|
| `enabled` | `boolean` | `false` | — | Whether data quantisation is active |
| `inputLevels` | `number` | `256` | `2..65536` | Number of discrete quantisation levels for input values |
| `outputLevels` | `number` | `0` | `0` or `2..65536` | Quantisation levels for target outputs (0 = disabled, preserves exact targets) |

### 2. Wire Config into the Configuration System

Following the [config pattern](https://github.com/stSoftwareAU/NEAT-AI/blob/Develop/src/config/CrossValidationConfig.ts):

1. Add `dataQuantisation: RequiredDataQuantisationConfig` to `NeatArguments.ts`
2. Add `dataQuantisation?: DataQuantisationConfig` to `NeatOptions.ts` (both `NeatOptions` and `NeatOptionsInput` types, with `CoerceNumeric<>` for CLI numeric fields)
3. Add `dataQuantisation` to both `Omit` lists in `NeatOptions.ts`
4. Parse in `NeatConfig.ts` / `NeatConfigParsers.ts` using `parseNumber()` with IIFE pattern
5. Pass through to `TrainOptions.ts` / `TrainArguments` so the training loop can access it

### 3. Implement Quantisation in Training Loop

In `src/architecture/Training.ts`, function `trainDirBinary()`:

**Integration point** — after copying record data into pre-allocated buffers (lines ~399-400) and **before** `activateAndTrace()`:

```typescript
// After copying into buffers, apply quantisation:
if (quantisationConfig.enabled) {
  quantiseBuffer(observationsBuffer, quantisationConfig.inputLevels);
  if (quantisationConfig.outputLevels > 0) {
    quantiseBuffer(targetsBuffer, quantisationConfig.outputLevels);
  }
}
```

Create a utility function (e.g., `src/propagate/DataQuantisation.ts` or similar) that quantises a `Float32Array` in-place:

```typescript
/**
 * Quantise values to a fixed number of discrete levels.
 * Maps the value range [min, max] observed in the buffer to `levels` evenly spaced bins.
 * Each value is snapped to the nearest bin centre.
 */
function quantiseBuffer(buffer: Float32Array, levels: number): void {
  // Find min/max in the buffer
  let min = buffer[0], max = buffer[0];
  for (let i = 1; i < buffer.length; i++) {
    if (buffer[i] < min) min = buffer[i];
    if (buffer[i] > max) max = buffer[i];
  }
  const range = max - min;
  if (range < Number.EPSILON) return; // All values identical, nothing to quantise

  const step = range / (levels - 1);
  for (let i = 0; i < buffer.length; i++) {
    buffer[i] = min + Math.round((buffer[i] - min) / step) * step;
  }
}
```

**Note**: Quantisation is deterministic (no randomness needed), but applied per-record to handle varying value ranges across records. Consider whether the min/max should be per-buffer (adaptive) or configured globally — per-buffer is simpler and more robust to start.

### 4. Extend to Predictive Coding Trainer

The `trainWithPredictiveCoding()` path in `PredictiveCodingTrainer.ts` has its own data reading loop. Apply the same quantisation there so that all training modes benefit.

### 5. Tests

Add tests in `test/` directory (mirroring `src/` structure):
- Unit tests for the quantisation function:
  - Verify output values are snapped to expected discrete levels
  - Verify edge cases: single value, all identical values, two values
  - Verify level count is respected (e.g., 4 levels produces at most 4 distinct values)
- Unit tests for config parsing (verify defaults, bounds, validation)
- Integration test: train a small network with quantisation enabled and verify it completes without error
- Verify that quantisation is NOT applied during inference (only during training)

## Acceptance Criteria

- [ ] `DataQuantisationConfig.ts` follows the established three-type config pattern
- [ ] Config is wired into `NeatArguments`, `NeatOptions`, and `NeatConfig`
- [ ] Quantisation reduces continuous values to the configured number of discrete levels
- [ ] Input and output quantisation can be controlled independently
- [ ] Disabled by default (backward compatible — no behaviour change unless opted in)
- [ ] Works with both standard backpropagation and predictive coding training paths
- [ ] Cross-validation path automatically inherits quantisation (via `trainDirBinary`)
- [ ] All tests pass, `./quality.sh` succeeds
- [ ] Australian English spelling throughout (e.g., "quantisation", "behaviour")

## Context

- Quantisation is a standard technique for preventing memorisation — by reducing precision, the network cannot learn exact input-output mappings
- Complement to fuzzing (#1900): fuzzing adds random noise while quantisation discretises values. Both can be used together
- Lower `levels` values = more aggressive quantisation = stronger memorisation prevention but potential loss of useful signal. 256 levels (≈8-bit precision) is a reasonable default that significantly reduces memorisation while preserving most information
- Unlike fuzzing, quantisation is deterministic — the same input always maps to the same quantised value

Part of #1899

## Notes
- Our production training data is generally normalised to -1...1 
- The quantisation must work on floats between these range. I was initially thing we would use float 16 or float 8 but if there's a more standard approach that's fine, I'll go with best practice. 
- I think this may need to be an initial Rust scan of the training data and quantise into a seperate training data directory so that the overhead is a one time per training run. 
- In production our training runs are over 10-20gb training data files. Speed is paramount when training.  
- I want to be able to pass an option from the calling program and everything under the hood is automatic.

**Comment by nleck** [comment]
Is input levels the normal approach when we're talking about observations which are floats between -1...1 ?

**Comment by stservice** [comment]
Claimed by `GRQ-23`

**Comment by stservice** [comment]
## Answer

Great question — and an important one to get right before implementation.

### Is "input levels" the standard approach for quantising floats in [-1, 1]?

Yes, **uniform quantisation with a fixed number of levels is the standard approach**, and it maps well to your [-1, 1] normalised data. Here's why:

**How it works with your data range:**

With `inputLevels: 256` and data in [-1, 1], the quantiser divides the range into 255 equal intervals (256 fence posts), giving a step size of `2 / 255 ≈ 0.00784`. Each float snaps to the nearest discrete level:

```
levels:  -1.0, -0.9922, -0.9843, ..., 0.0, ..., 0.9843, 0.9922, 1.0
step:    ≈0.00784
```

This is essentially the same as converting a float to 8-bit fixed-point and back — a well-established technique in neural network quantisation (used in TensorFlow Lite, PyTorch quantisation, ONNX Runtime, etc.).

**The issue description's per-buffer adaptive min/max approach vs fixed range:**

The issue proposes finding min/max per buffer and quantising within that observed range. Since your production data is already normalised to [-1, 1], you have two options:

| Approach | Pros | Cons |
|----------|------|------|
| **Fixed range [-1, 1]** | Deterministic across records, simpler, no min/max scan needed | Wastes levels if data only uses part of the range |
| **Per-buffer adaptive** (as proposed) | Maximises resolution for each record's actual range | Slightly more compute (min/max scan), different records quantise differently |

For your use case, I'd recommend **fixed range [-1, 1]** as the default since you already know the data range. It's simpler, faster (no min/max scan), and fully deterministic. You could add an `adaptive` boolean option if per-buffer behaviour is desired later.

**Regarding alternative approaches:**

- **Float16/Float8 casting**: This is a different kind of quantisation — it reduces storage precision but doesn't give you control over the number of discrete levels. Float16 gives ~3-4 decimal digits of precision (far more levels than 256), and bfloat16/float8 are primarily for GPU compute efficiency, not memorisation prevention. The "levels" approach gives you explicit, tuneable control over how much information is preserved.

- **Stochastic quantisation**: Instead of snapping to the nearest level, you probabilistically round up or down proportional to distance. This can be useful when combining with gradient-based training, but adds randomness (which you already have via fuzzing from #1900). Deterministic quantisation is the simpler and more standard starting point.

### Performance consideration from the notes

The suggestion to do a one-time Rust pre-processing pass into a separate directory is worth considering for 10-20GB datasets. However, since the quantisation function proposed here is simple arithmetic on contiguous Float32Arrays (no branching, cache-friendly), it should be very fast even in TypeScript — likely under 1ms per record. The existing data fuzzing (#1900) already does a similar per-element pass in the hot loop without issues.

If performance does become a concern, the fixed-range approach (rather than adaptive per-buffer) would be faster since it eliminates the min/max scan entirely — just a multiply, round, multiply per element:

```typescript
// Fixed range [-1, 1], no min/max scan needed
const step = 2 / (levels - 1);
for (let i = 0; i < buffer.length; i++) {
  buffer[i] = Math.round((buffer[i] + 1) / step) * step - 1;
}
```

### Interaction with fuzzing (#1900)

When both are enabled, the order matters. The current proposal applies quantisation first, then fuzzing would add noise on top. This is the correct order — quantise to reduce precision, then fuzz to add variability. The two techniques complement each other well: quantisation provides deterministic precision reduction while fuzzing adds stochastic perturbation.

### Short answer

Yes, "input levels" (uniform quantisation) is the standard and appropriate approach for [-1, 1] float data. 256 levels (≈8-bit precision) is a reasonable default that mirrors industry practice in neural network quantisation. Consider using a fixed [-1, 1] range rather than adaptive per-buffer min/max, since your data is already normalised — it's simpler, faster, and fully deterministic.

---

*This answer was generated automatically by the Vibe Coder. Add the `question` label again to ask follow-up questions.*
---

🤖 Processed by: stservice

**Comment by nleck** [comment]
We do know the range of values for our production market training data it's -1..1 

This is a generic library we can't assume that's true for other training data. I would like to be able to specify ( maybe default to the normalized range if enabled.  Only if I enable should anything change. 

Performance is key... it's already quite slow ( on 20gb of training data) so wouldn't want to make it worse

**Comment by stservice** [comment]
Claimed by `GRQ-25`

**Comment by stservice** [comment]
Claimed by `GRQ-23`

**Comment by stservice** [comment]
Pull request https://github.com/stSoftwareAU/NEAT-AI/pull/1907 has been created to address this issue.

**Comment by stservice** [comment]
Closed automatically — PR #1907 has been merged.
